drivers/modbus_rtu/include/modbus_rtu.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+/**
+ * @defgroup    drivers_modbus ModBus
+ * @ingroup     drivers
+ * @brief       Modbus
+ *
+ * @{
+ * @file
+ * @brief       Modbus interface definitions
+ *
+ * @author      Denis Litvinov <li.denis.iv@gmail.com>
+ */
+
+#include "periph/uart.h"
+#include "thread.h"
+#include "mutex.h"
+
+/**
+ * @brief Modbus functions
+ */
+enum MB_FC {
+  MB_FC_NONE = 0,                /**< null operator */
+  MB_FC_READ_COILS = 1,          /**< FCT=1 -> read coils or digital outputs */
+  MB_FC_READ_DISCRETE_INPUT = 2, /**< FCT=2 -> read digital inputs */
+  MB_FC_READ_REGISTERS = 3,      /**< FCT=3 -> read registers or analog outputs */
+  MB_FC_READ_INPUT_REGISTER = 4, /**< FCT=4 -> read analog inputs */
+  MB_FC_WRITE_COIL = 5,          /**< FCT=5 -> write single coil or output */
+  MB_FC_WRITE_REGISTER = 6,      /**< FCT=6 -> write single register */
+  MB_FC_WRITE_COILS = 15,        /**< FCT=15 -> write multiple coils or outputs */
+  MB_FC_WRITE_REGISTERS = 16     /**< FCT=16 -> write multiple registers */
+};
+
+//! Максимальный размер пакета протокола Modbus RTU.
+#define MODBUS_RTU_PACKET_SIZE_MAX 256
+
+//! Размер контрольной суммы протокола Modbus RTU.
+#define MODBUS_RTU_CRC_SIZE 2
+
+//! Размер полей сообщения протокола Modbus RTU.
+#define MODBUS_RTU_FIELDS_SIZE (sizeof(modbus_rtu_id_t) + sizeof(modbus_rtu_func_t))
+//! Размер полей и контрольной суммы сообщения протокола Modbus RTU.
+#define MODBUS_RTU_FIELDS_CRC_SIZE (MODBUS_RTU_CRC_SIZE + MODBUS_RTU_FIELDS_SIZE)
+//! Максимальный размер данных протокола Modbus RTU.
+#define MODBUS_RTU_DATA_SIZE_MAX (MODBUS_RTU_PACKET_SIZE_MAX - MODBUS_RTU_FIELDS_CRC_SIZE)
+//! Максимальный размер данных и контрольной суммы протокола Modbus RTU.
+#define MODBUS_RTU_DATA_CRC_SIZE_MAX (MODBUS_RTU_DATA_SIZE_MAX + MODBUS_RTU_CRC_SIZE)
+
+/**
+ * @brief Modbus message structure.
+ */
+typedef struct {
+  uint8_t id;     /**< ID of slave, on response not use */
+  uint8_t func;   /**< function code */
+  uint16_t addr;  /**< Starting Address */
+  uint16_t count; /**< Quantity of Registers or byte or bits */
+  uint16_t *regs; /**< Registers or byte or bits*/
+} modbus_rtu_message_t;
+
+/**
+ * @brief Modbus structure.
+ */
+typedef struct {
+  /**
+   * @brief  UART port.
+   */
+  uart_t uart;
+  /**
+   * @brief  baudrate.
+   */
+  uint32_t baudrate;
+  /**
+   * @brief  RTS pin. 0 - disable.
+   */
+  gpio_t pin_rts;
+  /**
+   * @brief  RTS signals when transmit.
+   */
+  int pin_rts_enable;
+  /**
+   * @brief  ID device. 0 - master, otherwice slave.
+   */
+  uint8_t id;
```

Comment:
You could safe a couple of bytes by reordering the members in the struct so that fewer padding bytes are needed. E.g. if you order members by alignment requirement, you would end up with as few padding bytes added as absolutely needed.

Reply:
``` C
typedef struct {
  ...
  modbus_rtu_message_t *msg;
  kernel_pid_t pid;
  uint8_t id;
  uint8_t size_buffer;
} modbus_rtu_t;
```
Means reorder like this?

Reply:
I have to see the whole struct to confirm this, but the part shown looks good.

The problem I'm referring to is best illustrated with an example. Let's say we have a structure `stuct foo1` declared like this:

```C
struct foo1 {
    uint8_t a;
    uint32_t b
    uint8_t c;
    uint8_t d;
    uint8_t e;
};
```

The whole structure has an alignment requirement of 4 bytes, as the maximum of the alignment requirements of its members is 4 bytes. (Member `uint32_t b` has an alignment requirement of 4 bytes, at least on 32 bit ARM CPUs. All other members have an alignment requirement of 1 bytes.) The C compiler will now fill padding bytes in front of every member whose padding requirement is not fulfilled, until it is fulfilled. Finally, padding bytes are added in the end (if needed), until the padding requirement of the whole struct is fulfilled. (This is needed e.g. to allow an array of the struct.)

The actual memory layout would look like this:

```C
struct foo1 {
    uint8_t a;             /* <-- Alignment requirement of 1 byte */
    uint8_t __padding1[3]; /* <-- Three bytes of padding inserted */
    uint32_t b;            /* <-- Alignment requirement of 4 bytes */
    uint8_t c;             /* <-- Alignment requirement of 1 byte */
    uint8_t d;             /* <-- Alignment requirement of 1 byte */
    uint8_t e;             /* <-- Alignment requirement of 1 byte */
    uint8_t __pading2[1];  /* <-- one byte of padding inserted */
};
```

The following `struct foo2` would need no padding at all (and safe 4 bytes because of this), even though it contains the same members:

```C
struct foo2 {
    uint32_t b;
    uint8_t a;
    uint8_t c;
    uint8_t d;
    uint8_t e;
}
```

Alternatively, `struct foo3` would also need no padding:

```C
struct foo3 {
    uint8_t a;
    uint8_t c;
    uint8_t d;
    uint8_t e;
    uint32_t b;
}
```

To sum up: If you order members by alignment requirements, you will always come up with the least amount of padding. Alternatively, you can carefully order members yourself and also end up with the least amount of padding bytes, but possible with a more "sensible" order regarding what the members express. You can use whatever approach is more appealing to you. It would however be nice, if padding could be reduced to a minimum to safe same RAM. (In the end, we will access members by name anyway, so the order is not so relevant to us, but it is relevant to the size of the `struct`.)

Reply:
```C
typedef struct {
  gpio_t pin_rts; // unsigned
  int pin_rts_enable;
  uint32_t timeout;
  uint32_t rx_timeout;
  modbus_rtu_message_t *msg;
  mutex_t mutex_buffer; // list_node_t*
  uart_t uart; // unsigned
  kernel_pid_t pid; // uint16_t
  uint8_t id;
  uint8_t size_buffer;
  uint8_t buffer[MODBUS_RTU_PACKET_SIZE_MAX];
} modbus_rtu_t;
```
Full structure